src/NumericalAlgorithms/LinearSolver/InnerProduct.hpp

```diff
@@ -0,0 +1,57 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+///\file
+/// Defines an inner product for the linear solver
+
+#pragma once
+
+#include <array>
+
+#include "DataStructures/Variables.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Blaze.hpp"
+#include "Utilities/ForceInline.hpp"
+
+namespace LinearSolver {
+
+/// \ingroup LinearSolverGroup
+/// Implementations of LinearSolver::inner_product.
+namespace InnerProductImpls {
+
+/// The inner product between any types that have a `dot` product
+template <typename Lhs, typename Rhs>
+struct InnerProductImpl {
+  static double apply(const Lhs& lhs, const Rhs& rhs) noexcept {
+    return dot(lhs, rhs);
+  }
+};
+
+/// The inner product between `Variables`
+template <typename LhsTagsList, typename RhsTagsList>
+struct InnerProductImpl<Variables<LhsTagsList>, Variables<RhsTagsList>> {
+  static double apply(const Variables<LhsTagsList>& lhs,
+                      const Variables<RhsTagsList>& rhs) noexcept {
+    const auto size = lhs.size();
+    ASSERT(size == rhs.size(),
+           "The Variables must be of the same size to take an inner product");
+    return ddot_(size, lhs.data(), 1, rhs.data(), 1);
+  }
+};
+
+}  // namespace InnerProductImpls
+
+/*!
+ * \ingroup LinearSolverGroup
+ * \brief The inner product defined by virtue of a linear operator \f$A\f$.
```

Comment:
I'm confused about what is going on here.  There's nothing in the code that references a linear operator.  This is just a straight dot product.

Reply:
You're right, the operator A defines another (related) inner product when it's symmetric, which is used in the conjugate gradient algorithm.